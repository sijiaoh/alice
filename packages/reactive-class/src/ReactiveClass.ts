export interface Store {
  onChange: () => void;
}

export interface ReactiveClassConstructorParams {
  store: Store;
}

export class ReactiveClass {
  constructor({ store }: ReactiveClassConstructorParams) {
    this.store = store;
    return new Proxy(this, {
      set: (target, p: keyof ReactiveClass, value: never) => {
        // eslint-disable-next-line no-param-reassign
        target[p] = value;
        this.store.onChange();
        return true;
      },
    });
  }

  private store: Store;
}
