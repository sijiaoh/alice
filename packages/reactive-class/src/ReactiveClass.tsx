/* eslint-disable react-hooks/rules-of-hooks */

import autobind from 'autobind-decorator';
import { produce } from 'immer';
import React, { useEffect, useState } from 'react';
import { Props } from 'react-utils';
import {
  atom,
  RecoilState,
  useRecoilCallback,
  RecoilValue,
  Loadable,
  useRecoilState,
  RecoilRoot,
  useRecoilTransactionObserver_UNSTABLE,
} from 'recoil';

@autobind
export class ReactiveClass<T> {
  static readonly notProvideError = 'ReactiveClass.Provider not provided.';
  static get?: <U>(recoilValue: RecoilValue<U>) => Loadable<U>;
  static set?: <U>(
    recoilVal: RecoilState<U>,
    valOrUpdater: U | ((currVal: U) => U)
  ) => void;

  static Component({ children }: Props) {
    const dummy = useRecoilCallback(({ snapshot, set }) => {
      ReactiveClass.get = snapshot.getLoadable;
      ReactiveClass.set = set;
      return () => {};
    });

    useRecoilTransactionObserver_UNSTABLE(({ snapshot }) => {
      ReactiveClass.get = snapshot.getLoadable;
    });

    const [initialized, setInitialized] = useState(false);
    useEffect(() => {
      // jest用呼び出し。
      dummy();
      setInitialized(true);
    }, [dummy]);

    return <>{initialized ? children : null}</>;
  }

  static Provider({ children }: Props) {
    const { Component } = ReactiveClass;
    return (
      <RecoilRoot>
        <Component>{children}</Component>
      </RecoilRoot>
    );
  }

  readonly state: RecoilState<T>;

  get data() {
    return this.getData();
  }
  set data(data: Readonly<T>) {
    this.setData(data);
  }

  constructor(key: string, defaultStore: Readonly<T>) {
    this.state = atom({ key, default: defaultStore });
  }

  getData() {
    if (!ReactiveClass.get) throw new Error(ReactiveClass.notProvideError);
    return ReactiveClass.get(this.state).getValue();
  }

  setData(state: Readonly<T>) {
    if (!ReactiveClass.set) throw new Error(ReactiveClass.notProvideError);
    ReactiveClass.set(this.state, state);
  }

  changeValue(callback: (state: T) => Readonly<T> | void) {
    this.setData(produce(this.getData(), callback));
  }

  useState() {
    return useRecoilState(this.state);
  }
}
