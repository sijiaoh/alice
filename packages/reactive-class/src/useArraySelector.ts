import { useState, useEffect } from 'react';
import { ReactiveClass } from './ReactiveClass';

export function useArraySelector<T1 extends ReactiveClass, R>(
  vs: [T1],
  callback: (vs: [T1]) => R
): void;
export function useArraySelector<
  T1 extends ReactiveClass,
  T2 extends ReactiveClass,
  R
>(vs: [T1, T2], callback: (vs: [T1, T2]) => R): void;
export function useArraySelector<
  T1 extends ReactiveClass,
  T2 extends ReactiveClass,
  T3 extends ReactiveClass,
  R
>(vs: [T1, T2, T3], callback: (vs: [T1, T2, T3]) => R): void;

// eslint-disable-next-line @typescript-eslint/no-explicit-any
export function useArraySelector<T>(arg1: any, arg2: (vs: any) => T) {
  const vs = arg1 as ReactiveClass[];
  const callback = arg2 as (vs: ReactiveClass[]) => T;

  const [value, setValue] = useState(callback(vs));

  useEffect(() => {
    const unsubscribes = vs.map((v) => {
      return v.subscribe(() => {
        setValue(callback(vs));
      });
    });
    return () => {
      unsubscribes.forEach((unsubscribe) => unsubscribe());
    };
  }, []);

  return value;
}
