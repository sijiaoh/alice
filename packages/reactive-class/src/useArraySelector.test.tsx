/* eslint-disable max-classes-per-file */

import React from 'react';
import { create, act, ReactTestRenderer } from 'react-test-renderer';
import { Subject } from 'rxjs';
import { ReactiveClass } from './ReactiveClass';
import { useArraySelector } from './useArraySelector';

describe('useArraySelector', () => {
  it('update render when reactive class changed', () => {
    class Some1 extends ReactiveClass {
      num = 1;
    }
    class Some2 extends ReactiveClass {
      str = '1';
    }
    class Some3 extends ReactiveClass {
      arr = [1];
    }
    const some1 = new Some1();
    const some2 = new Some2();
    const some3 = new Some3();

    const Component = ({ s1, s2, s3 }: { s1: Some1; s2: Some2; s3: Some3 }) => {
      const prop = useArraySelector([s1, s2, s3], ([ls1, ls2, ls3]) => {
        return [ls1.num, ls2.str, ls3.arr];
      });
      return <>{JSON.stringify(prop)}</>;
    };

    let component!: ReactTestRenderer;
    void act(() => {
      component = create(<Component s1={some1} s2={some2} s3={some3} />);
    });
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    expect(((some1 as any).subject as Subject<Some1>).observers.length).toBe(1);
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    expect(((some2 as any).subject as Subject<Some2>).observers.length).toBe(1);
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    expect(((some3 as any).subject as Subject<Some3>).observers.length).toBe(1);
    expect(component.toJSON()).toMatchSnapshot();

    void act(() => {
      some1.num = 2;
      some2.str = '2';
      some3.arr = [2];
    });
    expect(component.toJSON()).toMatchSnapshot();

    void act(() => {
      component.unmount();
    });
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    expect(((some1 as any).subject as Subject<Some1>).observers.length).toBe(0);
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    expect(((some2 as any).subject as Subject<Some2>).observers.length).toBe(0);
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    expect(((some3 as any).subject as Subject<Some3>).observers.length).toBe(0);
    expect(component.toJSON()).toMatchSnapshot();
  });
});
