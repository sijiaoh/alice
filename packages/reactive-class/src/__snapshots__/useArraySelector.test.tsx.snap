// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useArraySelector update render when reactive class changed 1`] = `"[1,\\"1\\",[1]]"`;

exports[`useArraySelector update render when reactive class changed 2`] = `"[2,\\"2\\",[2]]"`;

exports[`useArraySelector update render when reactive class changed 3`] = `null`;
